torch/testing/_internal/distributed/distributed_test.py

```diff
@@ -2501,6 +2510,60 @@ def test_DistributedDataParallel_non_default_stream(self):
                         msg=f"Expected gradient of {expected_grad} but got {avg} on rank {self.rank}",
                     )
 
+        def _test_desync_helper(self, tensor):
+            rank = self.rank
+
+            dist.destroy_process_group()
+
+            # Enable NCCL Async error handling
+            os.environ["NCCL_ASYNC_ERROR_HANDLING"] = "1"
+
+            # Reinitialize global process group with a timeout
+            dist.init_process_group(
+                init_method=INIT_METHOD,
+                backend=BACKEND,
+                world_size=int(os.environ["WORLD_SIZE"]),
+                rank=rank,
+                timeout=timedelta(seconds=2),
+            )
+
+            dist.all_reduce(tensor)
```

Comment:
Could it lead to flakiness if this allreduce doesn't complete in 2s?

Reply:
It actually won't because the async error handling mechanism will kick in and abort the collective if it hasn't completed. The only reason we don't explicitly introduce the desync in the first collective is so that this workload more closely mirrors a real training run, where some collective (usually not the first) is blocked on a previous GPU operation.